app/models/spree/gift_card_barcode.rb

```diff
@@ -0,0 +1,47 @@
+module Spree
+  class GiftCardBarcode < Spree::Base
+    acts_as_paranoid
+    validates_uniqueness_of :number, unless: :deleted_at
+
+    belongs_to :virtual_gift_card, class_name: 'Spree::VirtualGiftCard'
+    belongs_to :order, class_name: 'Spree::Order'
+
+    scope :unredeemed, -> { where(redeemed_at: nil) }
+
+    self.whitelisted_ransackable_attributes = %w[number]
+
+    def sold!(gift_card, order)
+      self.order = order
+      self.virtual_gift_card = gift_card
+      self.sold_at = Time.now
+
+      save
+    end
+
+    def redeem(user)
+      return if redeemed?
+      user.store_credits.create!(amount: amount, category: store_credit_category, currency: 'USD', created_by: user)
+      redeem_associated_gift_card
+      update_attributes(redeemed_at: Time.now)
+    end
+
+    def store_credit_category
+      Spree::StoreCreditCategory.gift_card
+    end
+
+    def redeemed?
```

Comment:
Same with this and the class method below. To me, these make more sense on the virtual gift card object.